### ISSUE TITLE
Nondeterministic output of cstdint vs. stdint.h

### ISSUE THREAD

**Issue by ptomato** [include-what-you-use/include-what-you-use#805]
IWYU seems to sometimes suggest including `<cstdint>` when it sees `uint64_t` in the code, and sometimes suggest `<stdint.h>`. I think `<stdint.h>` is probably correct if you're using `uint64_t` and not `std::uint64_t`, but the main issue is that the output isn't the same every time (and will even sometimes suggest `<cstdint>` if you already have `<stdint.h>` included)

See [this GitLab CI job](https://gitlab.gnome.org/GNOME/gjs/-/jobs/786839) where the suggestion of `#include <cstdint>  // for uint64_t` disappeared after retrying the job, on the exact same commit.

(May be similar to #226, but I don't think quite the same, since this problem involves nondeterminism.)

**Comment by kimgr** [comment]
@ptomato Interesting, indeed! Will see if I can squeeze out some time to debug this at some point.

**Comment by justusranvier** [comment]
I've seen cases where it suggests <sys/types.h> instead of <cstdint> for `std::uint8_t`.

I wish there was a command line option I could use to tell iwyu to always include the c++ version of headers and never the c versions for my project. Manually fixing up the suggestions is very tiresome and greatly reduces the value of the fix_includes script.

**Comment by i-ky** [comment]
> I wish there was a command line option I could use to tell iwyu to always include the c++ version of headers and never the c versions for my project. Manually fixing up the suggestions is very tiresome and greatly reduces the value of the fix_includes script.

In my experience this usually points to another problem. Perhaps, your code uses `::uint8_t` instead of `std::uint8_t`.

**Comment by justusranvier** [comment]
> Perhaps, your code uses `::uint8_t` instead of `std::uint8_t`.

No, it properly use std::uint8_t: https://github.com/Open-Transactions/opentxs/blob/d790bb2f0105900ed0bced44a17824998946731b/src/blockchain/client/Wallet.hpp#L232

0.13 didn't suggest anything for this line, but 0.14 suggests `sys/types.h`.

There's never a case in my project where I want to include sys/types.h, or time.h, or stdlib.h, or any of the other c headers with  c++11 replacements, but I haven't ever found a way to successfully communicate this to iwyu.

**Comment by kimgr** [comment]
@justusranvier I think issue #407 covers that, but I'm not sure it's related to the non-determinism between C and C++ headers.
The way to communicate it to IWYU is mappings, even if it's annoying to introduce mappings to work around bugs.

**Comment by kimgr** [comment]
Huh, turns out #407 was fixed for 0.14, so that's probably why you're seeing the report. The `sys/types.h` vs `<cstdint>` confusion is probably due to missing mappings for the `std::xxx_t` types from `<cstdint>`. 

Can you guys try with this:
```diff
commit fcdf390bd7a487230f1be863c5ce14c220ebd107
Author: Kim Grasman <kim.grasman@gmail.com>
Date:   Fri Jul 31 12:52:01 2020 +0200

    Add symbol mappings for <cstdint> types

diff --git a/gcc.symbols.imp b/gcc.symbols.imp
index 9f498d8..e7dc83a 100644
--- a/gcc.symbols.imp
+++ b/gcc.symbols.imp
@@ -165,5 +165,16 @@
   # string/ostream/istream type.
   { symbol: [ "std::char_traits", private, "<string>", public ] },
   { symbol: [ "std::char_traits", private, "<ostream>", public ] },
-  { symbol: [ "std::char_traits", private, "<istream>", public ] }
+  { symbol: [ "std::char_traits", private, "<istream>", public ] },
+  # Special stdint mappings for C++
+  { symbol: [ "std::int8_t", private, "<cstdint>", public ] },
+  { symbol: [ "std::int16_t", private, "<cstdint>", public ] },
+  { symbol: [ "std::int32_t", private, "<cstdint>", public ] },
+  { symbol: [ "std::int64_t", private, "<cstdint>", public ] },
+  { symbol: [ "std::uint8_t", private, "<cstdint>", public ] },
+  { symbol: [ "std::uint16_t", private, "<cstdint>", public ] },
+  { symbol: [ "std::uint32_t", private, "<cstdint>", public ] },
+  { symbol: [ "std::uint64_t", private, "<cstdint>", public ] },
+  { symbol: [ "std::intptr_t", private, "<cstdint>", public ] },
+  { symbol: [ "std::uintptr_t", private, "<cstdint>", public ] }
 ]
diff --git a/iwyu_include_picker.cc b/iwyu_include_picker.cc
index bbd49eb..0c0849b 100644
--- a/iwyu_include_picker.cc
+++ b/iwyu_include_picker.cc
@@ -248,6 +248,18 @@ const IncludeMapEntry libstdcpp_symbol_map[] = {
   { "std::char_traits", kPrivate, "<string>", kPublic },
   { "std::char_traits", kPrivate, "<ostream>", kPublic },
   { "std::char_traits", kPrivate, "<istream>", kPublic },
+  // Special stdint mappings for C++
+  { "std::int8_t", kPrivate, "<cstdint>", kPublic },
+  { "std::int8_t", kPrivate, "<cstdint>", kPublic },
+  { "std::int16_t", kPrivate, "<cstdint>", kPublic },
+  { "std::int32_t", kPrivate, "<cstdint>", kPublic },
+  { "std::int64_t", kPrivate, "<cstdint>", kPublic },
+  { "std::uint8_t", kPrivate, "<cstdint>", kPublic },
+  { "std::uint16_t", kPrivate, "<cstdint>", kPublic },
+  { "std::uint32_t", kPrivate, "<cstdint>", kPublic },
+  { "std::uint64_t", kPrivate, "<cstdint>", kPublic },
+  { "std::intptr_t", kPrivate, "<cstdint>", kPublic },
+  { "std::uintptr_t", kPrivate, "<cstdint>", kPublic },
 };
 
 // Private -> public include mappings for GNU libc
```

The non-determinism is probably something else, but maybe this is a start.

**Comment by ptomato** [comment]
> Can you guys try with this

Was that referring to the original problem I reported or the one from https://github.com/include-what-you-use/include-what-you-use/issues/805#issuecomment-665137783 ?

**Comment by kimgr** [comment]
@ptomato Sorry, no, this was for the other problem. It could be that it works aorund the non-determinism too, but that seems like a deeper problem that would be nice to get to the bottom of. Is it possible to come up with a reduced repro?

**Comment by ptomato** [comment]
I have seen it I think 2 or 3 times on CI runs, but I cannot seem to reproduce it locally (or at least not once in about 50 attempts)

if it's not reproducible enough then do feel free to close this! :smile:

**Comment by kimgr** [comment]
@ptomato Thanks for trying it out. I'll let it rest for a while and see if I can come up with anything.

**Comment by ptomato** [comment]
I have still not been able to reproduce the nondeterminism, but I added a minimal example to #226 that consistently reproduces the problem that I saw intermittently before.

**Comment by carlosgalvezp** [comment]
@kimgr I have tried your solution above and I still get suggestions to add `stdint.h` instead of `cstdint`. Did it work for you?

**Comment by kimgr** [comment]
@ptomato Thanks to #982, I think I finally have a clue here.

* The files to analyze are stored as a set, and therefore analyzed in unspecified order
* If there's only a single main file with a single associated header (e.g. `foo.cpp` with `foo.h`, which is the common case), the associated is analyzed first, and then the main, no surprises
* But if there are multiple associated headers, interdependencies between them can cause confusion
* Besides associated headers `--check_also` adds additional files to the set of files to analyze, and I suspect triggers the same kind of confusion

QED: https://gitlab.gnome.org/GNOME/gjs/-/blob/master/tools/run_iwyu.sh#L92 :smiley:

I still need to think about how/if/why there's interference between these additional files, but I think this might be signal that the root cause is in sight.

**Comment by ptomato** [comment]
That's great news, let me know if I can help you by testing anything.
